### PR TITLE
Add OLMoE

### PR DESCRIPTION
**Pull request by fizzAI**
yoinked from vllm PR but changed to (hopefully) work in aphro, i haven't actually tested

### CODE REVIEW

**Comment by fizzAI** [comment]
tested on 1xA40 and it works fine
don't have any benchmark script on me but it seems faster than either vllm or kcpp on the same hardware^-^

File: aphrodite/modeling/models/olmoe.py [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+        super().__init__()
+        self.hidden_size = hidden_size
+
+        # Gate always runs at half / full precision for now.
+        self.gate = ReplicatedLinear(hidden_size,
+                                     num_experts,
+                                     bias=False,
+                                     quant_config=None)
+
+        self.experts = FusedMoE(num_experts=num_experts,
```
Comment (AlpinDale) [inline aphrodite/modeling/models/olmoe.py:66]: Have you checked if this works with quants?

Comment (fizzAI) [reply aphrodite/modeling/models/olmoe.py:66]: no

Comment (fizzAI) [reply aphrodite/modeling/models/olmoe.py:66]: but i will, spinning up another pod

Comment (fizzAI) [reply aphrodite/modeling/models/olmoe.py:66]: well, with fp8, it throws a too old error; but it should work w/ marlin kernels, right? or are those model-specific
traceback:
<details>
<pre>
Process SpawnProcess-1:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/triton/language/core.py", line 35, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/triton/language/core.py", line 1597, in load
    return semantic.load(pointer, mask, other, boundary_check, padding_option, cache_modifier, eviction_policy,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/triton/language/semantic.py", line 1037, in load
    return _load_legacy(ptr, mask, other, boundary_check, padding, cache, eviction, is_volatile, builder)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/triton/language/semantic.py", line 1005, in _load_legacy
    other = cast(other, elt_ty, builder)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/triton/language/semantic.py", line 759, in cast
    assert builder.options.allow_fp8e4nv, "fp8e4nv data type is not supported on CUDA arch < 89"
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: fp8e4nv data type is not supported on CUDA arch < 89

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/endpoints/openai/rpc/server.py", line 209, in run_rpc_server
    server = AsyncEngineRPCServer(async_engine_args, rpc_path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/endpoints/openai/rpc/server.py", line 24, in __init__
    self.engine = AsyncAphrodite.from_engine_args(async_engine_args)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/engine/async_aphrodite.py", line 601, in from_engine_args
    engine = cls(
             ^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/engine/async_aphrodite.py", line 510, in __init__
    self.engine = self._init_engine(*args, **kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/engine/async_aphrodite.py", line 694, in _init_engine
    return engine_class(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/engine/aphrodite_engine.py", line 274, in __init__
    self._initialize_kv_caches()
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/engine/aphrodite_engine.py", line 336, in _initialize_kv_caches
    self.model_executor.determine_num_available_blocks())
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/executor/gpu_executor.py", line 94, in determine_num_available_blocks
    return self.driver_worker.determine_num_available_blocks()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/task_handler/worker.py", line 196, in determine_num_available_blocks
    self.model_runner.profile_run()
  File "/usr/local/lib/python3.11/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/task_handler/model_runner.py", line 1125, in profile_run
    self.execute_model(model_input, kv_caches, intermediate_tensors)
  File "/usr/local/lib/python3.11/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/task_handler/model_runner.py", line 1553, in execute_model
    hidden_or_intermediate_states = model_executable(
                                    ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/models/olmoe.py", line 308, in forward
    hidden_states = self.model(input_ids, positions, kv_caches,
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/models/olmoe.py", line 273, in forward
    hidden_states, residual = layer(positions, hidden_states,
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/models/olmoe.py", line 233, in forward
    hidden_states = self.mlp(hidden_states)
                    ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/models/olmoe.py", line 82, in forward
    final_hidden_states = self.experts(hidden_states=hidden_states,
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/layers/fused_moe/layer.py", line 280, in forward
    final_hidden_states = self.quant_method.apply(
                          ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/quantization/fp8.py", line 485, in apply
    return fused_experts(x,
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/layers/fused_moe/fused_moe.py", line 544, in fused_experts
    invoke_fused_moe_kernel(curr_hidden_states,
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/layers/fused_moe/fused_moe.py", line 258, in invoke_fused_moe_kernel
    fused_moe_kernel[grid](
  File "/usr/local/lib/python3.11/dist-packages/triton/runtime/jit.py", line 345, in <lambda>
    return lambda *args, **kwargs: self.run(grid=grid, warmup=False, *args, **kwargs)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/triton/runtime/jit.py", line 662, in run
    kernel = self.compile(
             ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/triton/compiler/compiler.py", line 276, in compile
    module = src.make_ir(options, codegen_fns, context)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/triton/compiler/compiler.py", line 113, in make_ir
    return ast_to_ttir(self.fn, self, context=context, options=options, codegen_fns=codegen_fns)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
triton.compiler.errors.CompilationError: at 119:12:

    # -----------------------------------------------------------
    # Iterate to compute a block of the C matrix.
    # We accumulate into a `[BLOCK_SIZE_M, BLOCK_SIZE_N]` block
    # of fp32 values for higher accuracy.
    # `accumulator` will be converted back to fp16 after the loop.
    accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.float32)

    for k in range(0, tl.cdiv(K, BLOCK_SIZE_K)):
        # Load the next block of A and B, generate a mask by checking the
        # K dimension.
        a = tl.load(a_ptrs,
            ^
</pre>
 </details>
 quant llm (and probably the other methods, but those arent on-runtime so i don't want to check) error with something like
 <details>
<pre>
Process SpawnProcess-1:
Traceback (most recent call last):
  File "/usr/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/endpoints/openai/rpc/server.py", line 209, in run_rpc_server
    server = AsyncEngineRPCServer(async_engine_args, rpc_path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/endpoints/openai/rpc/server.py", line 24, in __init__
    self.engine = AsyncAphrodite.from_engine_args(async_engine_args)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/engine/async_aphrodite.py", line 601, in from_engine_args
    engine = cls(
             ^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/engine/async_aphrodite.py", line 510, in __init__
    self.engine = self._init_engine(*args, **kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/engine/async_aphrodite.py", line 694, in _init_engine
    return engine_class(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/engine/aphrodite_engine.py", line 261, in __init__
    self.model_executor = executor_class(
                          ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/executor/executor_base.py", line 45, in __init__
    self._init_executor()
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/executor/gpu_executor.py", line 36, in _init_executor
    self.driver_worker.load_model()
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/task_handler/worker.py", line 153, in load_model
    self.model_runner.load_model()
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/task_handler/model_runner.py", line 888, in load_model
    self.model = get_model(model_config=self.model_config,
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/model_loader/__init__.py", line 20, in get_model
    return loader.load_model(model_config=model_config,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/model_loader/loader.py", line 337, in load_model
    model = _initialize_model(model_config, self.load_config,
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/model_loader/loader.py", line 166, in _initialize_model
    return build_model(
           ^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/model_loader/loader.py", line 151, in build_model
    return model_class(config=hf_config,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/models/olmoe.py", line 293, in __init__
    self.model = OlmoeModel(config, cache_config, quant_config)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/models/olmoe.py", line 253, in __init__
    self.layers = nn.ModuleList([
                                ^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/models/olmoe.py", line 254, in <listcomp>
    OlmoeDecoderLayer(config,
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/models/olmoe.py", line 197, in __init__
    self.mlp = OlmoeMoE(
               ^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/models/olmoe.py", line 66, in __init__
    self.experts = FusedMoE(num_experts=num_experts,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/aphrodite/modeling/layers/fused_moe/layer.py", line 185, in __init__
    assert self.quant_method is not None
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
</pre>
</details>

Comment (fizzAI) [reply aphrodite/modeling/models/olmoe.py:66]: to my eye, it looks like it really _should_ be properly passing `quant_method` into the model `__init__`? seems odd

Comment (AlpinDale) [reply aphrodite/modeling/models/olmoe.py:66]: As a workaround, you may want to port over the MoE implementation from mixtral_quant.py. You can have the MoE method selection be conditional on the presence of quant_method (or create an olmoe_quant.py, up to you). If quantization support isn't super important for this model arch, then we can merge as-is and worry about it later.

Comment (fizzAI) [reply aphrodite/modeling/models/olmoe.py:66]: i mean, the only olmoe model is a 7b/1b active atm, i'm not 100% sure it _needs_ quant support
personally i'd say merge now and i can PR quant support later (atm i'm doing this all between the github web editor and a runpod with nano, not the best dev environment)

**Review by AlpinDale — APPROVED** [review]

**Comment by AlpinDale** [comment]
Looks like the linter tests are failing, can you run `./formatting.sh`?

**Comment by fizzAI** [comment]
I'll run it when I get off of work in a few hours
Also, I just realized, is there a reason `transformers==4.44.1` in the requirements? That talk in kobo made me realize that to get this PR to work, I had to force upgrade transformers (which I did without thinking about it)

**Comment by AlpinDale** [comment]
> I'll run it when I get off of work in a few hours
> Also, I just realized, is there a reason `transformers==4.44.1` in the requirements? That talk in kobo made me realize that to get this PR to work, I had to force upgrade transformers (which I did without thinking about it)

We usually pin transformer versions for every new model release that requires it. You can bump it for this PR. We'd rather have pinned versions so every environment is reproducible and we don't run into annoying dependency hells.

**Comment by fizzAI** [comment]
Ya, I understand why it's pinned, was just wondering if there was something abt that specific version that kept it pinned

**Pull request merged**